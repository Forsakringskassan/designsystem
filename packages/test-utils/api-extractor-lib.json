{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/lib.d.ts",
    "apiReport": {
        "enabled": false
    },
    "dtsRollup": {
        "publicTrimmedFilePath": "<projectFolder>/dist/types/lib.d.ts"
    }
}
