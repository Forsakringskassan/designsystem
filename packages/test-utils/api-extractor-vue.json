{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/vue/index.d.ts",
    "apiReport": {
        "enabled": false
    },
    "dtsRollup": {
        "publicTrimmedFilePath": "<projectFolder>/dist/types/vue.d.ts"
    }
}
