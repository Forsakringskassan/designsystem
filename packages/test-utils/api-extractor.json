{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "tsdocMetadata": {
        "tsdocMetadataFilePath": "<projectFolder>/dist/types/tsdoc-metadata.json"
    },
    "dtsRollup": {
        "enabled": false
    }
}
