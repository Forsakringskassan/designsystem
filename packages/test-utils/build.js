const fs = require("node:fs/promises");
const path = require("node:path");
const esbuild = require("esbuild");

const type = {
    cjs: "commonjs",
    esm: "module",
};

async function build() {
    for (const format of ["cjs", "esm"]) {
        const outdir = `dist/${format}`;
        const common = {
            logLevel: "info",
            sourcemap: true,
            external: ["vue", "@fkui/*"],
            bundle: true,
            platform: "node",
            format,
            target: "node20",
            outdir,
        };
        await esbuild.build({
            ...common,
            entryPoints: ["src/lib.ts", "src/jest.ts"],
        });
        await esbuild.build({
            ...common,
            entryPoints: [{ in: "src/vue/index.ts", out: "vue" }],
            define: {
                "process.env.NODE_ENV": JSON.stringify("development"),
            },
        });
        const pkg = JSON.stringify({ type: type[format] }, null, 2);
        await fs.writeFile(path.join(outdir, "package.json"), pkg, "utf-8");
    }
}

build().catch((err) => {
    console.error(err);
});
