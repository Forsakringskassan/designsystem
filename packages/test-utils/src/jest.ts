/* imported for side-effects only */
import "./matchers";

export * from "./lib";
