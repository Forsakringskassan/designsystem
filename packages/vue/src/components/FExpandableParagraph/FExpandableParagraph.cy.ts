import { defineComponent } from "vue";
import {
    DensityWrapper,
    densityWrapperHeight,
    densityWrapperWidth,
} from "@fkui/test-utils/vue";
import FExpandableParagraph from "./FExpandableParagraph.vue";

const TestComponent = defineComponent({
    template: /* HTML */ `
        <f-expandable-paragraph :expanded>
            <template #title> Rubrik </template>
            <template #default> Inneh√•ll </template>
        </f-expandable-paragraph>
    `,
    props: {
        expanded: {
            type: Boolean,
            default: false,
        },
    },
    components: {
        FExpandableParagraph,
    },
});

describe("density", () => {
    const DensityComponent = defineComponent({
        template: /* HTML */ `
            <density-wrapper>
            <test-component expanded></test-component
            </density-wrapper>
        `,
        components: {
            DensityWrapper,
            TestComponent,
        },
    });

    it(`should be densified`, () => {
        cy.viewport(densityWrapperWidth, densityWrapperHeight);
        cy.mount(DensityComponent);
        cy.toMatchScreenshot();
    });
});

it("should match visual regression when collapsed", () => {
    cy.mount({
        template: /* HTML */ `
            <div id="wrapper" style="padding: 1rem">
                <div style="border: 1px dashed hotpink">
                    <test-component></test-component>
                </div>
            </div>
        `,
        components: {
            TestComponent,
        },
    });
    cy.get("#wrapper").toMatchScreenshot();
});

it("should match visual regression when expanded", () => {
    cy.mount({
        template: /* HTML */ `
            <div id="wrapper" style="padding: 1rem">
                <div style="border: 1px dashed hotpink">
                    <test-component expanded></test-component>
                </div>
            </div>
        `,
        components: {
            TestComponent,
        },
    });
    cy.get("#wrapper").toMatchScreenshot();
});
