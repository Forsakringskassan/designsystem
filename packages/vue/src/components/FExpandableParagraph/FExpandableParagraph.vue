<script lang="ts">
import { defineComponent } from "vue";
import { ElementIdService } from "@fkui/logic";
import { FIcon } from "../FIcon";
import { hasSlot } from "../../utils";
import { FExpand } from "../FExpand";

export default defineComponent({
    name: "FExpandableParagraph",
    components: {
        FIcon,
        FExpand,
    },
    inheritAttrs: false,
    props: {
        /**
         * Toggle expanded/collapsed state
         */
        expanded: {
            type: Boolean,
            required: false,
            default: false,
        },
        /**
         * Element to render for the button elmement
         * 'span', 'h1', 'h2', 'h3', 'h4', 'h5' and 'h6' is valid.
         */
        headerTag: {
            default: "span",
            required: false,
            validator(value: string): boolean {
                return ["h1", "h2", "h3", "h4", "h5", "h6", "span"].includes(value);
            },
        },
        /**
         * Visual header size for the button element.
         * 'h1', 'h2', 'h3', 'h4', 'h5' and 'h6' is valid.
         * If not used, the default styling will be h4
         */
        headerVisualTag: {
            type: String,
            default: "",
            required: false,
            validator(value: string): boolean {
                return ["", "h1", "h2", "h3", "h4", "h5", "h6"].includes(value);
            },
        },
        /**
         * Toggle list style of component
         */
        list: {
            type: Boolean,
            default: false,
            required: false,
        },
        /**
         * The id for the content id attribute.
         * If the prop is not set the id will be generated.
         */
        id: {
            type: String,
            required: false,
            default: () => ElementIdService.generateElementId(),
        },
    },
    emits: ["toggle"],
    computed: {
        expandedClass(): string {
            return this.expanded ? "expandable-paragraph--open" : "expandable-paragraph--closed";
        },
        listClass(): string {
            return this.list ? "expandable-paragraph--list" : "";
        },
        relatedClass(): string {
            return this.hasRelatedSlot ? "expandable-paragraph--related-information" : "";
        },
        hasRelatedSlot(): boolean {
            return hasSlot(this, "related");
        },
        headerVisualClass(): string {
            return `heading--${this.headerVisualTag || "h4"}`;
        },
    },
    methods: {
        onClickMinimize(event: MouseEvent): void {
            /**
             * Emitted when heading is clicked.
             * @type {MouseEvent}
             */
            this.$emit("toggle", event);
        },
    },
});
</script>

<template>
    <div class="expandable-paragraph" :class="[expandedClass, listClass]">
        <div :class="relatedClass">
            <component :is="headerTag" class="expandable-paragraph__heading" :class="headerVisualClass">
                <button
                    type="button"
                    class="expandable-paragraph__button"
                    :aria-expanded="expanded ? 'true' : 'false'"
                    :aria-controls="id"
                    v-bind="$attrs"
                    @click="onClickMinimize"
                >
                    <span class="expandable-paragraph__icon">
                        <span class="icon-stack">
                            <f-icon name="dash"></f-icon>
                            <f-icon name="dash"></f-icon>
                        </span>
                    </span>
                    <!-- @slot Slot used for title content -->
                    <slot name="title"></slot>
                </button>
            </component>
            <div v-if="hasRelatedSlot" class="expandable-paragraph__related-information">
                <!-- @slot Slot used for related information besides the header element -->
                <slot name="related"></slot>
            </div>
        </div>
        <f-expand>
            <div v-show="expanded" :id="id" class="expandable-paragraph__container">
                <div class="expandable-paragraph__content">
                    <!-- @slot Slot used for content shown when expanded -->
                    <slot></slot>
                </div>
                <div v-if="!list" class="expandable-paragraph__separator"></div>
            </div>
        </f-expand>
    </div>
</template>
