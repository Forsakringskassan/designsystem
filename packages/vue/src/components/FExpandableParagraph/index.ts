export { default as FExpandableParagraph } from "./FExpandableParagraph.vue";
