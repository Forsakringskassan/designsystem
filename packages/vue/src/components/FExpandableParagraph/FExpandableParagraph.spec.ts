import { VueWrapper, mount, shallowMount } from "@vue/test-utils";
import { type ConfigData } from "html-validate";
import FExpandableParagraph from "./FExpandableParagraph.vue";
import "html-validate/jest";

const config: ConfigData = {
    root: true,
    rules: {
        "no-inline-style": "off",
    },
};

function createWrapper({
    props = {},
    slots = {},
    attrs = {},
} = {}): VueWrapper {
    return mount(FExpandableParagraph, {
        attrs: { ...attrs },
        props: { id: "my-id", ...props },
        slots: {
            title: "My expandable component",
            default: "Lorem ipsum",
            ...slots,
        },
        global: {
            stubs: ["f-expand", "f-icon"],
        },
    });
}

describe("snapshots", () => {
    it("should match snapshot when collapsed", () => {
        expect.assertions(2);
        const wrapper = createWrapper();
        expect(wrapper.element).toMatchSnapshot();
        expect(wrapper.element).toHTMLValidate(config);
    });

    it("should match snapshot when expanded", () => {
        expect.assertions(2);
        const wrapper = createWrapper({ props: { expanded: true } });
        expect(wrapper.element).toMatchSnapshot();
        expect(wrapper.element).toHTMLValidate(config);
    });

    it("should match snapshot with custom heading level", () => {
        expect.assertions(2);
        const wrapper = createWrapper({ props: { headerTag: "h3" } });
        expect(wrapper.element).toMatchSnapshot();
        expect(wrapper.element).toHTMLValidate(config);
    });

    it("should match snapshot with list styling", () => {
        expect.assertions(2);
        const wrapper = createWrapper({
            props: { headerTag: "h3", list: true },
        });
        expect(wrapper.element).toMatchSnapshot();
        expect(wrapper.element).toHTMLValidate(config);
    });

    it("should match snapshot with related information", () => {
        expect.assertions(2);
        const wrapper = createWrapper({
            props: { headerTag: "h3" },
            slots: { related: "dolor sit amet" },
        });
        expect(wrapper.element).toMatchSnapshot();
        expect(wrapper.element).toHTMLValidate(config);
    });
});

describe("header-visual-tag", () => {
    it("should default to h4 visual heading", () => {
        expect.assertions(1);
        const wrapper = shallowMount(FExpandableParagraph, {});
        const heading = wrapper.get(".expandable-paragraph__heading");
        expect(heading.classes()).toEqual([
            "expandable-paragraph__heading",
            "heading--h4",
        ]);
    });

    it("should change the visual heading class when used", () => {
        expect.assertions(1);
        const wrapper = shallowMount(FExpandableParagraph, {
            props: { headerVisualTag: "h6" },
        });
        const heading = wrapper.get(".expandable-paragraph__heading");
        expect(heading.classes()).toEqual([
            "expandable-paragraph__heading",
            "heading--h6",
        ]);
    });
});

describe("events", () => {
    it("should emit toggle event", async () => {
        expect.assertions(1);
        const toggle = jest.fn();

        const wrapper = createWrapper({
            attrs: { onToggle: toggle },
        });
        const input = wrapper.get("button");
        await input.trigger("click");

        expect(toggle).toHaveBeenCalled();
    });

    it("should pass listeners", async () => {
        const foobar = jest.fn();
        const wrapper = createWrapper({
            attrs: { onFoobar: foobar },
        });
        const element = wrapper.get("button");
        await element.trigger("foobar");
        expect(foobar).toHaveBeenCalled();
    });
});
