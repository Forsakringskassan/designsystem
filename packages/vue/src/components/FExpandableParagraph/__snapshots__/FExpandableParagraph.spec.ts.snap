// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots should match snapshot when collapsed 1`] = `
<div
  class="expandable-paragraph expandable-paragraph--closed"
>
  <div
    class=""
  >
    <span
      class="expandable-paragraph__heading heading--h4"
    >
      <button
        aria-controls="my-id"
        aria-expanded="false"
        class="expandable-paragraph__button"
        type="button"
      >
        <span
          class="expandable-paragraph__icon"
        >
          <span
            class="icon-stack"
          >
            <f-icon-stub
              library="f"
              name="dash"
            />
            <f-icon-stub
              library="f"
              name="dash"
            />
          </span>
        </span>
        
        My expandable component
        
      </button>
    </span>
    <!--v-if-->
  </div>
  <f-expand-stub />
</div>
`;

exports[`snapshots should match snapshot when expanded 1`] = `
<div
  class="expandable-paragraph expandable-paragraph--open"
>
  <div
    class=""
  >
    <span
      class="expandable-paragraph__heading heading--h4"
    >
      <button
        aria-controls="my-id"
        aria-expanded="true"
        class="expandable-paragraph__button"
        type="button"
      >
        <span
          class="expandable-paragraph__icon"
        >
          <span
            class="icon-stack"
          >
            <f-icon-stub
              library="f"
              name="dash"
            />
            <f-icon-stub
              library="f"
              name="dash"
            />
          </span>
        </span>
        
        My expandable component
        
      </button>
    </span>
    <!--v-if-->
  </div>
  <f-expand-stub />
</div>
`;

exports[`snapshots should match snapshot with custom heading level 1`] = `
<div
  class="expandable-paragraph expandable-paragraph--closed"
>
  <div
    class=""
  >
    <h3
      class="expandable-paragraph__heading heading--h4"
    >
      <button
        aria-controls="my-id"
        aria-expanded="false"
        class="expandable-paragraph__button"
        type="button"
      >
        <span
          class="expandable-paragraph__icon"
        >
          <span
            class="icon-stack"
          >
            <f-icon-stub
              library="f"
              name="dash"
            />
            <f-icon-stub
              library="f"
              name="dash"
            />
          </span>
        </span>
        
        My expandable component
        
      </button>
    </h3>
    <!--v-if-->
  </div>
  <f-expand-stub />
</div>
`;

exports[`snapshots should match snapshot with list styling 1`] = `
<div
  class="expandable-paragraph expandable-paragraph--closed expandable-paragraph--list"
>
  <div
    class=""
  >
    <h3
      class="expandable-paragraph__heading heading--h4"
    >
      <button
        aria-controls="my-id"
        aria-expanded="false"
        class="expandable-paragraph__button"
        type="button"
      >
        <span
          class="expandable-paragraph__icon"
        >
          <span
            class="icon-stack"
          >
            <f-icon-stub
              library="f"
              name="dash"
            />
            <f-icon-stub
              library="f"
              name="dash"
            />
          </span>
        </span>
        
        My expandable component
        
      </button>
    </h3>
    <!--v-if-->
  </div>
  <f-expand-stub />
</div>
`;

exports[`snapshots should match snapshot with related information 1`] = `
<div
  class="expandable-paragraph expandable-paragraph--closed"
>
  <div
    class="expandable-paragraph--related-information"
  >
    <h3
      class="expandable-paragraph__heading heading--h4"
    >
      <button
        aria-controls="my-id"
        aria-expanded="false"
        class="expandable-paragraph__button"
        type="button"
      >
        <span
          class="expandable-paragraph__icon"
        >
          <span
            class="icon-stack"
          >
            <f-icon-stub
              library="f"
              name="dash"
            />
            <f-icon-stub
              library="f"
              name="dash"
            />
          </span>
        </span>
        
        My expandable component
        
      </button>
    </h3>
    <div
      class="expandable-paragraph__related-information"
    >
      
      dolor sit amet
      
    </div>
  </div>
  <f-expand-stub />
</div>
`;
